File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea, stack = 0, []
        heights.append(0)
        for idx, height in enumerate(heights):
            start = idx
            while stack and stack[-1][1] > height:
                cur, h = stack.pop()
                maxArea = max(maxArea, h * (idx - cur))
                start = cur
                
            stack.append((start, height))
            
        return maxArea

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0);
        int n = heights.size(), maxArea = 0, start = 0;
        stack<pair<int,int>> st;
        
        for (int i = 0; i < n; i++) {
            start = i;
            while (!st.empty() && st.top().second > heights[i]) {
                start = st.top().first;
                maxArea = max(maxArea, st.top().second * (i - start));
                st.pop();
            }
            st.push({start, heights[i]});
        }
        return maxArea;
    }
};